### ISSUE TITLE
UpSampling2D not supported in Layer , float16 issue  

### ISSUE THREAD

**Issue by ArikEv** [ARM-software/armnn#533]
Hi,

I have two issues :

1. Does the ARMNN supports UpSampling2D in layer of UNet  ?
2. I use the tflite version 2.5.0 , in addition I use PyArmNN. I tried to load my tflite model in PyArmNN
    (CreateNetworkFromBinaryFile("tflite")) .
NN for example  :

    down1 = Conv2D(4, (3, 3), padding='same', )(input_shape)
    down1 = Activation('relu')(down1)
    down1_pool = MaxPooling2D((2, 2), strides=(2, 2))(down1)

    When I use float16 I get the following error:
    `Failed to parse operator #2 within subgraph #0 error: Buffer #0 has 0 bytes. For tensor: [4,3,3,3] expecting: 432 bytes and 
     108 elements.  at function CreateConstTensor`
     But when I use  float32 I have no any errors

BR,
Arik.

**Comment by matthewsloyanARM** [comment]
Hi @ArikEv, 

Thank you for reaching out. To reply to your questions.

1. UpSampling2D is not supported currently by ArmNN, however as UpSampling2D is used simply to scale up an image using nearest neighbor or bilinear methods, it could be achieved using the ResizeBilinear or ResizeNearestNeighbor operators which are supported in ArmNN.
2. It is quite possible an operator in your model doesn't have float16 support in the specific backend you are using. Would you be able to supply the model if possible? Also, what backend are you using (E.g. CpuRef, CpuAcc or GpuAcc)?

Thank you, all the best!

Kind regards,

Matthew

**Comment by ArikEv** [comment]
Hi Matthew,

Thank you for fast response!!!

1. I will try to use the ResizeBilinear or ResizeNearestNeighbor operators.
2. The source that I use in order to create TFLite file 
```
def saveTFLite_FP16(model):
  converter = tf.lite.TFLiteConverter.from_keras_model(model)
  converter.optimizations = [tf.lite.Optimize.DEFAULT]
  converter.target_spec.supported_types = [tf.float16]

  tfliteModel     = converter.convert()
  tfliteModelFile = pathlib.Path("./")/"model_fp16_armnn.tflite"
  print(f"Saving to TFLite FP16 v2 - {tfliteModelFile}")
  tfliteModelFile.write_bytes(tfliteModel)
```

```
model = keras.Sequential([
  keras.layers.InputLayer(input_shape=(28, 28, 3)),
  keras.layers.Conv2D(filters=12, kernel_size=(3, 3), activation=tf.nn.relu),
  keras.layers.MaxPooling2D(pool_size=(2, 2)),
  keras.layers.Flatten(),
  keras.layers.Dense(10)
])
```

The source that I use in order to activate ARMNN

```
  parser  = ann.ITfLiteParser()
  network = parser.CreateNetworkFromBinaryFile(["tflite"])

  options   = ann.CreationOptions()
  runTime   = ann.IRuntime(options)
  backends  = [ ann.BackendId('CpuAcc'), ann.BackendId('CpuRef')]

  optNetwork,_  = ann.Optimize(network, backends, runTime.GetDeviceSpec(), ann.OptimizerOptions())
  netID,_       = runTime.LoadNetwork(optNetwork)

  inputNames        = parser.GetSubgraphInputTensorNames(netID)
  inputBindingInfo  = parser.GetNetworkInputBindingInfo(netID, inputNames[0])
  inputs            = ann.make_input_tensors([inputBindingInfo], [image])

  outputNames       = parser.GetSubgraphOutputTensorNames(netID)
  outputBindingInfo = parser.GetNetworkOutputBindingInfo(netID, outputNames[0])
  outputs           = ann.make_output_tensors([outputBindingInfo])

  timeElapsed       = np.zeros(nIter)

  runTime.EnqueueWorkload(0, inputs, outputs) # Run inference
```
BR,

Arik.

**Comment by matthewsloyanARM** [comment]
Hi @ArikEv, 

Thank you for supplying the code. As you are running on CpuAcc float16 might not be supported on your CPU. What device/architecture are you running this on?

In the meantime I will go about trying to recreate this issue with the code above.

Kind regards,

Matthew

**Comment by ArikEv** [comment]
Hi Matthew,

ARM based server , Architecture-aarch64 , Family-ARMv8 , Model name-Neoverse-N1.
Flags: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs

BR,
Arik.

**Comment by matthewsloyanARM** [comment]
Hi @ArikEv, 

Thank you for supplying this information. I have created the fp16 and fp32 model using the code you supplied and what seems to be the issue is that Weights and Biases are used as inputs to the Conv2d layer. This is not supported in Arm NN yet. It is currently on our backlog but there is no ETA at the moment.

Here is the fp16 model visualized using Netron and below it is the fp32 model. It looks like the way TensorFlow handles the fp16 optimization creates this difference so that's why the fp32 model will work with ArmNN.
![image](https://user-images.githubusercontent.com/70519899/114852556-c6487200-9dda-11eb-96a3-fd37d49be802.png)
![image](https://user-images.githubusercontent.com/70519899/114853299-8cc43680-9ddb-11eb-85a6-bc9ed2475525.png)

Kind regards,

Matthew

**Comment by ArikEv** [comment]
Hi Matthew,

Thank you so much for the support,

BR,
Arik.
